### PR TITLE
Add deterministic path for CUDA `cumsum`

### LINKED ISSUE #75240: Large cumulative sums appear to be nondeterministic.
### 🐛 Describe the bug

@timothytmzhou and I found that cumulative sums over large input arrays do not seem to deterministically produce the same output array:

```python
import torch
import math
import random
import numpy as np
import os

# make ourselves as deterministic as possible.
os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
np.random.seed(0)
torch.manual_seed(0)
random.seed(0)
torch.use_deterministic_algorithms(True)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

# (deterministic) messy number generator
v = torch.Tensor([(x*math.pi)%1 for x in range(1000000)]).to(torch.device("cuda"))

# init a and b to the same cumulative sum.
a = b = v.cumsum(0)
counter = 0

# this loop should never terminate.
while ((a==b).all()):
	b = v.cumsum(0)
	counter+=1

print(f"non-determinism after {counter} runs.")
print("errors:",(a-b).abs()[a-b!=0])
print("first error index:",torch.where((a-b)!=0)[0][0])
```
yields, on my machine:
```
non-determinism after 1 runs.
errors: tensor([0.0020, 0.0020, 0.0020,  ..., 0.2188, 0.2188, 0.2188], device='cuda:0')
first error index: tensor(38400, device='cuda:0')
```
Multiple runs of this file exhibit unique non-determinism. Here's another run:
```
non-determinism after 1 runs.
errors: tensor([0.0039, 0.0039, 0.0039,  ..., 0.0938, 0.0938, 0.0938], device='cuda:0')
first error index: tensor(76800, device='cuda:0')
```
We decided to test this after reading [this paper](https://www.cs.purdue.edu/homes/lintan/publications/variance-ase20.pdf). Of note is the following paragraph:

> Scheduling and floating-point precision: GPU programming uses warp as a unit of computation. A warp consists of 32 parallel threads with concurrent memory access. Due to the limited precision (32-bits) of DL models, rounding errors are introduced at every step of floating-point calculation. In the GPU model, concurrent accesses to a single memory address must be serialized to prevent race conditions with no guaranteed order of access. Therefore, the rounding error introduced in each warp may vary across fixed-seed identical training runs due to the different accessing orders.

### Versions

Collecting environment information...
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 11.2.0
Clang version: 13.0.1
CMake version: version 3.22.3
Libc version: glibc-2.35

Python version: 3.10.4 (main, Mar 23 2022, 23:05:40) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.16.16-arch1-1-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.6.112
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1050 Ti
Nvidia driver version: 510.54
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.3.2
/usr/lib/libcudnn_adv_infer.so.8.3.2
/usr/lib/libcudnn_adv_train.so.8.3.2
/usr/lib/libcudnn_cnn_infer.so.8.3.2
/usr/lib/libcudnn_cnn_train.so.8.3.2
/usr/lib/libcudnn_ops_infer.so.8.3.2
/usr/lib/libcudnn_ops_train.so.8.3.2
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.3
[pip3] torch==1.11.0
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @ngimel @mruberry @kurtamohler

### LINKED ISSUE #89492: Feature Request: deterministic CUDA cumsum
### 🚀 The feature, motivation and pitch

There is no deterministic implementation for `cumsum_cuda_kernel`. It would be nice to have this!

### Alternatives

_No response_

### Additional context

This can currently still be used by setting the `warn_only` flag to `True`.
Warning:
`cumsum_cuda_kernel does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at ../aten/src/ATen/Context.cpp:82.)`

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @ptrblck @mruberry @kurtamohler @ngimel

### PR DESCRIPTION
**Pull request by kurtamohler**
Change `cumsum` to call its decomposition when `use_deterministic_algorithms(True)` and input is CUDA.

Fixes #89492
Fixes #75240

cc @mruberry

### CODE REVIEW

File: torch/__init__.py [hunk 1]
```diff
@@ -2587,6 +2587,27 @@ def get_device_module(device: _Optional[_Union[torch.device, str]] = None):
     return device_module
 
 
+def cumsum(
```
Comment (kurtamohler) [inline torch/__init__.py:2590]: I'm not sure exactly what is the right way to override the Python `torch.cumsum` function. This way causes a few mypy failures

Comment (kurtamohler) [reply torch/__init__.py:2590]: I realized that I need to put this in `torch/functional.py`

File: torch/functional.py [hunk 2]
```diff
@@ -2035,6 +2038,56 @@ def chain_matmul(*matrices, out=None):
...
+        raise RuntimeError("expected either 'dim' or 'axis' to be given, not both")
+    if axis is not None:
+        dim = axis
+    if has_torch_function_unary(self):
+        return handle_torch_function(cumsum, (self,), self, dim, dtype=dtype, out=out)
+    if not torch.jit.is_scripting():
+        if torch.are_deterministic_algorithms_enabled() and self.is_cuda:
+            ref_func = importlib.import_module("torch._refs").cumsum
+            return ref_func(self, dim, dtype=dtype, out=out)
+    return _VF.cumsum(self, dim, dtype=dtype, out=out)
```
Comment (kurtamohler) [inline torch/functional.py:2088]: I'm getting this mypy error:

```
Error (MYPY) [attr-defined]
  Module has no attribute "cumsum"

  >>> 2088  |    return _VF.cumsum(self, dim, dtype=dtype, out=out)
```

I'm not sure why it thinks `_VF.cumsum` doesn't exist. I guess I need to find out where type hints for `_VF` are defined and fix it

Comment (kurtamohler) [reply torch/functional.py:2088]: Ah, looks like the other functions in here use `  # type: ignore[attr-defined]` to ignore this

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
test fail looks real

**Comment by kurtamohler** [comment]
> test fail looks real

I'm having trouble figuring out how to fix that

File: test/onnx/test_operators.py [hunk 3]
```diff
@@ -873,7 +873,7 @@ def test_std(self):
 
     def test_cumsum(self):
         x = torch.randn(2, 3, 4, requires_grad=True)
-        self.assertONNX(lambda x: torch.cumsum(x, dim=1), x, opset_version=11)
+        self.assertONNX(lambda *args: torch.cumsum(*args, dim=1), x, opset_version=11)
```
Comment (kurtamohler) [inline test/onnx/test_operators.py:876]: This change fixes the failing test mentioned in https://github.com/pytorch/pytorch/pull/136224#issuecomment-2361912953

Apparently moving `torch.cumsum` to `torch/functional.py` caused the generated graph to pick up the name of the lambda arg "%x". I found that if I change the lambda arg to `my_arg`, for instance, the graph arg would be changed to "%my_arg" to match it.

So I figured I'd try using an args list instead--that way the generated graph isn't given a name for the arg, so it has to just use "%0"

Comment (justinchuby) [reply test/onnx/test_operators.py:876]: Feel free to do whatever that makes it work. I have a open PR to remove this test file completely but it is not in yet.

**Review by justinchuby — APPROVED** [review]

**Comment by atalman** [comment]
@kurtamohler  Looks like internal Changes are required to land this PR.

File: torch/functional.py [hunk 4]
```diff
@@ -2035,6 +2038,56 @@ def chain_matmul(*matrices, out=None):
...
+        raise RuntimeError("expected either 'dim' or 'axis' to be given, not both")
+    if axis is not None:
+        dim = axis
+    if has_torch_function_unary(self):
+        return handle_torch_function(cumsum, (self,), self, dim, dtype=dtype, out=out)
+    if not torch.jit.is_scripting():
+        if torch.are_deterministic_algorithms_enabled() and self.is_cuda:
+            ref_func = importlib.import_module("torch._refs").cumsum
+            return ref_func(self, dim, dtype=dtype, out=out)
+    return _VF.cumsum(self, dim, dtype=dtype, out=out)  # type: ignore[attr-defined]
```
Comment (ezyang) [inline torch/functional.py:2088]: This breaks FC in a funny way: the out overload, strictly speaking, does not actually take None. So what you have to do is test if out is None and if it is None, call `_VF.cumsum(self, dim, dtype=dtype`. Can you make that change?

Comment (kurtamohler) [reply torch/functional.py:2088]: Ah ok, yep I'll do that

File: torch/_tensor.py [hunk 5]
```diff
@@ -846,6 +846,31 @@ def symeig(self, eigenvectors=False):
...
+            return handle_torch_function(
+                Tensor.cumsum,
+                (self,),
+                self,
+                dim,
+                dtype=dtype,
+                out=out,
+                axis=axis,
+            )
+        return torch.cumsum(self, dim, dtype=dtype, out=out, axis=axis)
```
Comment (ezyang) [inline torch/_tensor.py:872]: Hmm, I guess we need to do it here too?

Comment (kurtamohler) [reply torch/_tensor.py]: Done. Also, I figured we should avoid giving both axis and dim

**Comment by ezyang** [comment]
third time is a charm

**Comment by atalman** [comment]
Following error:
```
RuntimeError: 
Arguments for call are not valid.
The following variants are available:
  
  aten::cumsum(Tensor self, int dim, *, ScalarType? dtype=None) -> Tensor:
  Keyword argument out unknown.
  
  aten::cumsum.dimname(Tensor self, str dim, *, ScalarType? dtype=None) -> Tensor:
  Expected a value of type 'str' for argument 'dim' but instead found type 'int'.
  
  aten::cumsum.dimname_out(Tensor self, str dim, *, ScalarType? dtype=None, Tensor(a!) out) -> Tensor(a!):
  Expected a value of type 'str' for argument 'dim' but instead found type 'int'.
  
  aten::cumsum.out(Tensor self, int dim, *, ScalarType? dtype=None, Tensor(a!) out) -> Tensor(a!):
  Expected a value of type 'Tensor' for argument 'out' but instead found type 'NoneType'.
```

**Comment by ezyang** [comment]
Whelp. I guess we have to put the logic in C++. Maybe we should just port the Python decomp to C++ then.

**Comment by kurtamohler** [comment]
> Whelp. I guess we have to put the logic in C++. Maybe we should just port the Python decomp to C++ then.

Sounds good, working on it

**Comment by kurtamohler** [comment]
I've ported the decomp to C++

**Review by eqy — APPROVED** [review]

File: aten/src/ATen/native/cuda/ScanKernels.cpp [hunk 6]
```diff
@@ -88,14 +91,44 @@ Tensor _logcumsumexp_cuda(const Tensor& self, int64_t dim) {
...
+
+// This function is ported from the cumsum decomp in `torch/_refs`.
+Tensor& cumsum_deterministic(Tensor& result, Tensor& self, int64_t dim) {
+  int64_t ndim = self.dim();
+  dim = canonicalize_dim(ndim, dim);
+  if (ndim == 0) {
+    return at::sum_out(result, self.unsqueeze(0), /*dim=*/IntArrayRef{0});
   }
+  self = self.unsqueeze(dim + 1);
+  Tensor rg = at::arange(self.size(dim), c10::TensorOptions().device(self.device()));
```
[the remark this answers is not in the document]
Comment (kurtamohler) [reply aten/src/ATen/native/cuda/ScanKernels.cpp:115]: Oh good point, thanks for bringing that up

[the remark this answers is not in the document]
Comment (xw285cornell) [reply aten/src/ATen/native/cuda/ScanKernels.cpp:115]: yeah this is making the deterministic model quite unsable imo, e.g. 

```
torch.cumsum(e = torch.rand([1, 1, 128256], device="cuda"), dim=-1)
```
This is using 60GB memory.

[the remark this answers is not in the document]
Comment (ngimel) [reply aten/src/ATen/native/cuda/ScanKernels.cpp:115]: The alternative is hard error in deterministic mode, so the current state is still better? Or would you prefer this PR reverted @xw285cornell ?

[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/native/cuda/ScanKernels.cpp:115]: There is often a funny situation where we buggily gave nondeterministic results but people prefer that over OOM/error 🤣

[the remark this answers is not in the document]
Comment (ngimel) [reply aten/src/ATen/native/cuda/ScanKernels.cpp:115]: the difference is only in non-deterministic warn mode, where previously we would warn and take non-deterministic path and now will oom. For non-deterministic error mode if someone preferred non-deterministic result, that's on them, error means error.

[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/native/cuda/ScanKernels.cpp:115]: It looks like some internal peeps are complaining, so I'm gonna yank this and we can talk about it more

[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/native/cuda/ScanKernels.cpp:115]: So.... what should we do? Is there a torch.compile version that we can ship instead?

**Comment by kurtamohler** [comment]
Replaced by #140887

**Pull request closed without merging**